crates/cdk/src/wallet/send.rs

```diff
@@ -45,135 +29,351 @@ impl Wallet {
-                let selected_proofs_amount = selected_proofs.total_amount()?;
-                let amount_to_send = match include_fees {
-                    true => amount + self.get_proofs_fee(&selected_proofs).await?,
-                    false => amount,
-                };
-                if selected_proofs_amount - amount_to_send <= *tolerance {
-                    selected_proofs
-                } else {
-                    tracing::info!("Could not select proofs while offline. Attempting swap");
-                    self.swap_from_unspent(amount, conditions, include_fees)
-                        .await?
-                }
-            }
+        // Include token memo
+        let send_memo = send.options.memo.or(memo);
+        let memo = send_memo.and_then(|m| if m.include_memo { Some(m.memo) } else { None });
 
-            // Handle all other cases where selection fails
-            (
-                SendKind::OfflineExact
-                | SendKind::OnlineExact
-                | SendKind::OfflineTolerance(_)
-                | SendKind::OnlineTolerance(_),
-                Err(_),
-                _,
-            ) => {
-                tracing::debug!("Could not select proofs");
-                return Err(Error::InsufficientFunds);
-            }
-        };
+        // Create and return token
+        Ok(Token::new(
+            self.mint_url.clone(),
+            proofs_to_send,
+            memo,
+            self.unit.clone(),
+        ))
+    }
+
+    /// Cancel prepared send
+    pub async fn cancel_send(&self, send: PreparedSend) -> Result<(), Error> {
+        tracing::info!("Cancelling prepared send");
+
+        self.localstore
+            .update_proofs_state(send.proofs().ys()?, State::Unspent)
+            .await?;
+
+        Ok(())
+    }
+}
+
+/// Prepared send
+pub struct PreparedSend {
+    amount: Amount,
+    options: SendOptions,
+    proofs_to_swap: Proofs,
+    swap_fee: Amount,
+    proofs_to_send: Proofs,
+    send_fee: Amount,
+}
+
+impl PreparedSend {
+    /// Amount
+    pub fn amount(&self) -> Amount {
+        self.amount
+    }
 
-        self.send_proofs(memo, send_proofs).await
+    /// Send options
+    pub fn options(&self) -> &SendOptions {
+        &self.options
     }
+
+    /// Proofs to swap
+    pub fn proofs_to_swap(&self) -> &Proofs {
+        &self.proofs_to_swap
+    }
+
+    /// Swap fee
+    pub fn swap_fee(&self) -> Amount {
+        self.swap_fee
+    }
+
+    /// Proofs to send
+    pub fn proofs_to_send(&self) -> &Proofs {
+        &self.proofs_to_send
+    }
+
+    /// Send fee
+    pub fn send_fee(&self) -> Amount {
+        self.send_fee
+    }
+
+    /// All proofs
+    pub fn proofs(&self) -> Proofs {
+        let mut proofs = self.proofs_to_swap.clone();
+        proofs.extend(self.proofs_to_send.clone());
+        proofs
+    }
+
+    /// Total fee
+    pub fn fee(&self) -> Amount {
+        self.swap_fee + self.send_fee
```

Comment:
Shouldn't this check the options?

```suggestion
        match self.options.include_fee {
            true => self.swap_fee + self.send_fee,
            false => self.swap_fee,
        }
```

Reply:
The send_fee should always be zero if include_fee == false. I think it's safer to keep as is.